File: app/src/main/java/net/ginteam/carmen/presenter/filter/FiltersPresenter.java
package net.ginteam.carmen.presenter.filter;

import net.ginteam.carmen.contract.filter.FilterContract;
import net.ginteam.carmen.manager.FiltersViewStateManager;
import net.ginteam.carmen.model.Pagination;
import net.ginteam.carmen.model.filter.FilterModel;
import net.ginteam.carmen.provider.ModelCallback;
import net.ginteam.carmen.provider.filter.FiltersProvider;
import net.ginteam.carmen.view.custom.FilterEditText;

import java.util.List;

/**
 * Created by Eugene on 1/15/17.
 */

public class FiltersPresenter implements FilterContract.Presenter {

    private FilterContract.View mView;

    private FiltersViewStateManager mFiltersViewStateManager;

    @Override
    public void fetchFiltersForCategory(int categoryId) {
        mView.showLoading(true);

        FiltersProvider
                .getInstance()
                .fetchForCategory(categoryId, new ModelCallback<List<FilterModel>>() {
                    @Override
                    public void onSuccess(List<FilterModel> resultModel) {
                        mView.showLoading(false);
                        mView.showFilters(resultModel);
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void updateResultsWithFilter(int categoryId, String filter) {
        mView.showLoading(true);

        FiltersProvider
                .getInstance()
                .calculateCount(categoryId, filter, new ModelCallback<Pagination>() {
                    @Override
                    public void onSuccess(Pagination resultModel) {
                        mView.showLoading(false);
                        mView.showResultsCount(resultModel.getTotalItemsCount());
                    }

                    @Override
                    public void onFailure(String message) {
                        mView.showLoading(false);
                        mView.showError(message);
                    }
                });
    }

    @Override
    public void saveViewState(int categoryId, List<FilterEditText> filters, String filterQuery) {
        mFiltersViewStateManager.saveFiltersState(categoryId, filters, filterQuery);
    }

    @Override
    public String restoreViewState(int categoryId, List<FilterEditText> filters) {
        mFiltersViewStateManager.restoreFiltersState(categoryId, filters);
        return mFiltersViewStateManager.restoreFiltersQuery();
    }

    @Override
    public void attachView(FilterContract.View view) {
        mView = view;
        mFiltersViewStateManager = FiltersViewStateManager.getInstance();
    }

    @Override
    public void detachView() {
        mView = null;
    }

}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/presenter/filter/FiltersPresenter.kt
package net.ginteam.carmen.kotlin.presenter.filter

import android.util.Log
import net.ginteam.carmen.kotlin.api.response.MetaSubscriber
import net.ginteam.carmen.kotlin.contract.FiltersContract
import net.ginteam.carmen.kotlin.manager.PreferencesManager
import net.ginteam.carmen.kotlin.manager.SharedPreferencesManager
import net.ginteam.carmen.kotlin.model.CategoryModel
import net.ginteam.carmen.kotlin.model.CompanyModel
import net.ginteam.carmen.kotlin.model.PaginationModel
import net.ginteam.carmen.kotlin.presenter.BasePresenter
import net.ginteam.carmen.kotlin.provider.FiltersDataSourceProvider
import net.ginteam.carmen.kotlin.provider.OfflineFiltersDataSourceProvider
import net.ginteam.carmen.kotlin.view.activity.filter.FiltersActivityViewState

/**
 * Created by eugene_shcherbinock on 2/20/17.
 */
class FiltersPresenter : BasePresenter <FiltersContract.View>(), FiltersContract.Presenter {

    private val mPreferences: PreferencesManager = SharedPreferencesManager
    private val mFiltersDataSourceProvider: FiltersDataSourceProvider = OfflineFiltersDataSourceProvider()

    override fun fetchFilters(category: CategoryModel) {
        mView?.showLoading(true)

        mFiltersDataSourceProvider
                .fetchFiltersForCategory(category.id)
                .subscribe({
                    Log.d("FiltersPresenter", "Filters count ${it.size}")
                    mView?.showLoading(false)
                    mView?.showFilters(it)
                })
    }

    override fun updateResultsCount(category: CategoryModel, filterQuery: String) {
        mView?.showCounterLoading(true)

        val filter = filterQuery.plus("city_id:${mPreferences.userCityModel!!.id}")
        mFiltersDataSourceProvider
                .fetchCompaniesCountWithParameters(category.id, filter)
                .subscribe(object : MetaSubscriber <List <CompanyModel>>() {
                    override fun success(model: List<CompanyModel>, pagination: PaginationModel) {
                        mView?.updateFilterResultsCount(pagination.totalItems)
                    }

                    override fun error(message: String, isNetworkError: Boolean) {
                        mView?.showError(message, isNetworkError)
                    }
                })
    }

    override fun saveCurrentViewState() {
        mView?.let {
            val currentViewState: FiltersContract.ViewState = it.getCurrentViewState()
            val viewStateSaver: FiltersContract.ViewState = FiltersActivityViewState

            with(currentViewState) {
                viewStateSaver.categoryId = categoryId
                viewStateSaver.filterQuery = filterQuery
                viewStateSaver.fieldsList = fieldsList
            }
        }
    }

    override fun tryToRestoreViewState(category: CategoryModel): FiltersContract.ViewState? {
        val filtersViewState: FiltersContract.ViewState = FiltersActivityViewState
        if (category.id == filtersViewState.categoryId) {
            return filtersViewState
        }
        return null
    }
}